arangod/GeneralServer/H2CommTask.cpp

```diff
@@ -0,0 +1,654 @@
+    response.setHeaderNCIfNotSet(StaticStrings::XContentTypeOptions, StaticStrings::NoSniff);
+  }
+
+  // append write buffer and statistics
+  double const totalTime = RequestStatistics::ELAPSED_SINCE_READ_START(stat);
+
+  // and give some request information
+  LOG_TOPIC("924cc", DEBUG, Logger::REQUESTS)
+      << "\"h2-request-end\",\"" << (void*)this << "\",\""
+      << this->_connectionInfo.clientAddress
+      << "\",\""
+      //      << GeneralRequest::translateMethod(::llhttpToRequestType(&_parser))
+      << "\",\"" << static_cast<int>(response.responseCode()) << "\","
+      << Logger::FIXED(totalTime, 6);
+
+  strm->response.reset(static_cast<HttpResponse*>(baseRes.release()));
+
+  int32_t mid = static_cast<int32_t>(strm->response->messageId());
+  while (!_waitingResponses.push(mid)) {
+    std::this_thread::yield();
+  }
+
+  signalWrite();  // starts writing if necessary
+}
+
+// queue the response onto the session, call only on IO thread
+template <SocketType T>
+void H2CommTask<T>::queueHttp2Responses() {
+  int32_t streamId;
+  while (_waitingResponses.pop(streamId)) {
+    Stream* strm = findStream(streamId);
+    if (strm == nullptr) {
+      LOG_TOPIC("e2773", ERR, Logger::REQUESTS)
+          << "response with message id " << streamId << "has not H2 stream";
+      TRI_ASSERT(false);
+      return;
+    }
+
+    auto& res = *strm->response;
+
+    // we need a continous block of memory for headers
+    std::vector<nghttp2_nv> nva;
+    nva.reserve(4 + res.headers().size());
+
+    std::string status = std::to_string(static_cast<int>(res.responseCode()));
+    nva.push_back({(uint8_t*)":status", (uint8_t*)status.data(), 7,
+                   status.size(), NGHTTP2_NV_FLAG_NO_COPY_NAME});
+
+    bool seenServerHeader = false;
+    // bool seenConnectionHeader = false;
+    for (auto const& it : res.headers()) {
+      std::string const& key = it.first;
+      std::string const& val = it.second;
+
+      // ignore content-length
+      if (key == StaticStrings::ContentLength || key == StaticStrings::Connection ||
+          key == StaticStrings::TransferEncoding) {
+        continue;
+      }
+
+      if (key == StaticStrings::Server) {
+        seenServerHeader = true;
+      }
+
+      nva.push_back({(uint8_t*)key.data(), (uint8_t*)val.data(), key.size(), val.size(),
+                     NGHTTP2_NV_FLAG_NO_COPY_NAME | NGHTTP2_NV_FLAG_NO_COPY_VALUE});
+    }
+    
+    // add "Server" response header
+    if (!seenServerHeader && !HttpResponse::HIDE_PRODUCT_HEADER) {
+      nva.push_back({(uint8_t*)"server", (uint8_t*)"ArangoDB", 6, 8,
+                     NGHTTP2_NV_FLAG_NO_COPY_NAME | NGHTTP2_NV_FLAG_NO_COPY_VALUE});
+    }
+
+    for (std::string const& cookie : res.cookies()) {
+      nva.push_back({(uint8_t*)"set-cookie", (uint8_t*)cookie.c_str(), 10,
+                     cookie.length(),
+                     NGHTTP2_NV_FLAG_NO_COPY_NAME | NGHTTP2_NV_FLAG_NO_COPY_VALUE});
```

Comment:
don't we need to the value here? or is `res.cookies()` guaranteed to be valid later on?

Reply:
no copies lives in the HttpResponse object, which in turn lives as long as the stream exists